jib-plugins-common/src/main/java/com/google/cloud/tools/jib/plugins/common/UpdateChecker.java

```diff
@@ -0,0 +1,191 @@
+        executorService.submit(
+            () ->
+                performUpdateCheck(
+                    skip,
+                    Preconditions.checkNotNull(ProjectInfo.VERSION),
+                    versionUrl,
+                    getConfigDir()));
+    executorService.shutdown();
+    return new UpdateChecker(messageFuture);
+  }
+
+  @VisibleForTesting
+  static Optional<String> performUpdateCheck(
+      boolean skip, String currentVersion, String versionUrl, Path configDir) {
+    // Abort if offline or update checks are disabled
+    if (skip || Boolean.getBoolean(PropertyNames.DISABLE_UPDATE_CHECKS)) {
+      return Optional.empty();
+    }
+
+    try {
+      Path configFile = configDir.resolve("config.json");
+      if (Files.exists(configFile)) {
+        // Abort if update checks are disabled
+        ConfigJsonTemplate config =
+            JsonTemplateMapper.readJsonFromFile(configFile, ConfigJsonTemplate.class);
+        if (config.disableUpdateCheck) {
+          return Optional.empty();
+        }
+      } else {
+        // Generate config file if it doesn't exist
+        ConfigJsonTemplate config = new ConfigJsonTemplate();
+        Files.createDirectories(configDir);
+        JsonTemplateMapper.writeTo(config, Files.newOutputStream(configFile));
+      }
+
+      // Check time of last update check
+      Path lastUpdateCheck = configDir.resolve("lastUpdateCheck");
+      if (Files.exists(lastUpdateCheck)) {
+        String fileContents =
+            new String(Files.readAllBytes(lastUpdateCheck), StandardCharsets.UTF_8);
+        try {
+          Instant modifiedTime = Instant.parse(fileContents);
+          if (modifiedTime.plus(Duration.ofDays(1)).isAfter(Instant.now())) {
+            return Optional.empty();
+          }
+        } catch (DateTimeParseException ex) {
+          // Ignore parse failure; assume update check hasn't been performed
+        }
+      }
+
+      // Check for update
+      HttpURLConnection connection = (HttpURLConnection) new URL(versionUrl).openConnection();
+      connection.setConnectTimeout(3000);
+      BufferedReader bufferedReader =
+          new BufferedReader(
+              new InputStreamReader(connection.getInputStream(), StandardCharsets.UTF_8));
+      String latestVersion = bufferedReader.readLine().trim();
+      connection.disconnect();
+
+      Files.write(lastUpdateCheck, Instant.now().toString().getBytes(StandardCharsets.UTF_8));
+      if (currentVersion.equals(latestVersion)) {
+        return Optional.empty();
+      }
+      return Optional.of(
+          "A new version of Jib ("
+              + latestVersion
+              + ") is available (currently using "
+              + currentVersion
+              + "). Update your build configuration to use the latest features and fixes!");
+
+    } catch (IOException ex) {
+      try {
+        Files.deleteIfExists(configDir.resolve("config.json"));
+        Files.deleteIfExists(configDir.resolve("lastUpdateCheck"));
```

Comment:
nit: `Files.deleteIfExists(configFile)` and `Files.deleteIfExists(lastUpdateCheck)` after putting these various outside the `try` block. As this method is very long, I think it helps avoid the mistake of using different path.